File: sdk/rust/src/metrics.rs
//! # 监控指标模块
//!
//! 这个模块提供了 `Metrics` 结构体，用于获取沙箱的资源使用情况和运行状态。
//! 通过监控指标，你可以：
//!
//! - 了解沙箱的资源消耗（CPU、内存、磁盘）
//! - 判断沙箱是否正在运行
//! - 优化资源分配和成本
//! - 诊断性能问题
//!
//! ## 监控指标说明
//!
//! | 指标 | 类型 | 说明 |
//! |------|------|------|
//! | `cpu_usage` | `f32` | CPU 使用率（百分比，0-100） |
//! | `memory_usage` | `u64` | 内存使用量（MiB） |
//! | `disk_usage` | `u64` | 磁盘使用量（字节） |
//! | `running` | `bool` | 沙箱是否正在运行 |
//!
//! ## 使用示例
//!
//! ```rust,no_run
//! use microsandbox_sdk::{PythonSandbox, BaseSandbox};
//!
//! #[tokio::main]
//! async fn main() -> Result<(), Box<dyn std::error::Error>> {
//!     let mut sandbox = PythonSandbox::create("test").await?;
//!     sandbox.start(None).await?;
//!
//!     // 获取监控接口
//!     let metrics = sandbox.metrics().await?;
//!
//!     // 检查沙箱是否运行
//!     if metrics.is_running().await? {
//!         println!("沙箱正在运行");
//!     }
//!
//!     // 获取 CPU 使用率
//!     if let Some(cpu) = metrics.cpu().await? {
//!         println!("CPU 使用率：{:.1}%", cpu);
//!     }
//!
//!     // 获取内存使用量
//!     if let Some(memory) = metrics.memory().await? {
//!         println!("内存使用：{} MiB", memory);
//!     }
//!
//!     // 获取磁盘使用量
//!     if let Some(disk) = metrics.disk().await? {
//!         println!("磁盘使用：{:.2} MB", disk as f64 / 1024.0 / 1024.0);
//!     }
//!
//!     // 获取所有指标
//!     let all = metrics.all().await?;
//!     println!("完整指标：{:?}", all);
//!
//!     sandbox.stop().await?;
//!     Ok(())
//! }
//! ```

use std::{error::Error, sync::Arc};

use serde_json::json;
use tokio::sync::Mutex;
use uuid::Uuid;

use crate::base::SandboxBase;

/// # 监控指标接口
///
/// `Metrics` 结构体提供了获取沙箱资源使用情况和运行状态的接口。
/// 它通过向 Microsandbox 服务器发送查询请求来获取实时指标。
///
/// ## 架构说明
///
/// ```text
/// ┌─────────────────┐         HTTP 请求         ┌─────────────────┐
/// │   Metrics       │ ───────────────────────►  │  Microsandbox   │
/// │                 │                          │     Server      │
/// │ - cpu()         │ ◄───────────────────────  │                 │
/// │ - memory()      │         JSON 响应          │ - 收集指标数据  │
/// │ - disk()        │                          │ - 返回指标      │
/// │ - is_running()  │                          │                 │
/// │ - all()         │                          │                 │
/// └─────────────────┘                          └─────────────────┘
/// ```
///
/// ## 设计说明
///
/// `Metrics` 使用 `Arc<Mutex<SandboxBase>>` 来共享对底层沙箱的访问，
/// 这与 `Command` 结构体的设计一致。
///
/// ## 字段说明
pub struct Metrics {
    /// 基础沙箱实例的共享引用
    ///
    /// 通过这个引用来：
    /// - 检查沙箱是否已启动
    /// - 获取服务器 URL 和认证信息
    /// - 发送指标查询请求
    base: Arc<Mutex<SandboxBase>>,
}

impl Metrics {
    /// # 创建新的 Metrics 实例
    ///
    /// 这是一个内部方法，通常通过沙箱的 `metrics()` 方法获取 `Metrics` 实例。
    ///
    /// ## 参数
    ///
    /// * `base` - 沙箱的共享引用
    ///
    /// ## 返回值
    ///
    /// 返回一个新的 `Metrics` 实例。
    pub fn new(base: Arc<Mutex<SandboxBase>>) -> Self {
        Self { base }
    }

    /// # 获取沙箱指标数据
    ///
    /// 这是一个内部方法，负责向服务器发送指标查询请求。
    /// 所有公共方法（`cpu()`、`memory()`、`disk()` 等）都使用这个方法。
    ///
    /// ## 请求流程
    ///
    /// ```text
    /// 1. 检查沙箱是否已启动
    ///    └── 未启动 → 返回 NotStarted 错误
    ///
    /// 2. 获取沙箱连接信息
    ///    - server_url: 服务器地址
    ///    - sandbox_name: 沙箱名称
    ///    - api_key: API 密钥（如果有）
    ///
    /// 3. 构建 JSON-RPC 请求
    ///    {
    ///      "jsonrpc": "2.0",
    ///      "method": "sandbox.metrics.get",
    ///      "params": {
    ///        "sandbox": "sandbox-name"
    ///      },
    ///      "id": "uuid-here"
    ///    }
    ///
    /// 4. 发送 HTTP POST 请求
    ///
    /// 5. 解析响应
    ///    ├── 检查 HTTP 状态码
    ///    ├── 检查 JSON-RPC 错误
    ///    └── 提取沙箱指标数据
    ///
    /// 6. 返回指标 JSON
    /// ```
    ///
    /// ## 错误处理
    ///
    /// 可能返回的错误：
    /// - [`SandboxError::NotStarted`] - 沙箱未启动
    /// - [`SandboxError::RequestFailed`] - 请求失败
    /// - [`SandboxError::InvalidResponse`] - 响应格式错误
    ///
    /// ## 返回值
    ///
    /// * `Ok(serde_json::Value)` - 沙箱指标数据
    /// * `Err(...)` - 获取失败
    ///
    /// ## 返回数据结构
    ///
    /// 服务器返回的指标数据格式：
    ///
    /// ```json
    /// {
    ///   "name": "sandbox-name",
    ///   "running": true,
    ///   "cpu_usage": 12.5,
    ///   "memory_usage": 256,
    ///   "disk_usage": 1048576
    /// }
    /// ```
    async fn get_metrics(&self) -> Result<serde_json::Value, Box<dyn Error + Send + Sync>> {
        // 检查沙箱是否已启动
        let is_started = {
            let base = self.base.lock().await;
            base.is_started
        };

        // 如果沙箱未启动，返回错误
        if !is_started {
            return Err(Box::new(crate::SandboxError::NotStarted));
        }

        // 提取沙箱连接信息
        // 使用块作用域来获取锁，锁会在块结束时自动释放
        let (server_url, sandbox_name, api_key) = {
            let base = self.base.lock().await;
            (
                base.server_url.clone(),
                base.name.clone(),
                base.api_key.clone(),
            )
        };

        // 构建 JSON-RPC 2.0 请求 payload
        // 使用 UUID v4 生成唯一的请求 ID
        let request_id = Uuid::new_v4().to_string();
        let payload = json!({
            "jsonrpc": "2.0",
            "method": "sandbox.metrics.get",
            "params": {
                "sandbox": sandbox_name,
            },
            "id": request_id,
        });

        // 创建 HTTP 客户端
        // 注意：这里每次都创建新的客户端，可以考虑优化为复用客户端
        let client = reqwest::Client::new();
        let mut req_builder = client
            .post(&format!("{}/api/v1/rpc", server_url))
            .json(&payload)
            .header("Content-Type", "application/json");

        // 如果有 API 密钥，添加认证头
        if let Some(key) = api_key {
            req_builder = req_builder.header("Authorization", format!("Bearer {}", key));
        }

        // 发送请求并处理错误
        let response = req_builder
            .send()
            .await
            .map_err(|e| Box::new(crate::SandboxError::RequestFailed(e.to_string())))?;

        // 检查 HTTP 响应状态码
        if !response.status().is_success() {
            let status = response.status();
            let error_text = response
                .text()
                .await
                .unwrap_or_else(|_| "Unknown error".to_string());
            return Err(Box::new(crate::SandboxError::RequestFailed(format!(
                "Failed to get sandbox metrics: {} - {}",
                status, error_text
            ))));
        }

        // 解析 JSON 响应
        let response_data: serde_json::Value = response
            .json()
            .await
            .map_err(|e| Box::new(crate::SandboxError::InvalidResponse(e.to_string())))?;

        // 检查 JSON-RPC 错误字段
        if let Some(error) = response_data.get("error") {
            let message = error
                .get("message")
                .and_then(|m| m.as_str())
                .unwrap_or("Unknown error");
            return Err(Box::new(crate::SandboxError::RequestFailed(format!(
                "Failed to get sandbox metrics: {}",
                message
            ))));
        }

        // 提取 result 字段
        let result = response_data.get("result").ok_or_else(|| {
            crate::SandboxError::InvalidResponse("Missing 'result' field".to_string())
        })?;

        // 提取 sandboxes 数组
        // 服务器可能返回多个沙箱的指标，我们只关心自己的
        let sandboxes = result
            .get("sandboxes")
            .and_then(|s| s.as_array())
            .ok_or_else(|| {
                crate::SandboxError::InvalidResponse("Missing 'sandboxes' array".to_string())
            })?;

        // 处理空响应（沙箱可能已被销毁）
        if sandboxes.is_empty() {
            return Ok(json!({}));
        }

        // 返回第一个（也是唯一一个）沙箱的数据
        // 在响应中，我们的沙箱应该在第一个位置
        Ok(sandboxes[0].clone())
    }

    /// # 获取所有指标
    ///
    /// 返回沙箱的完整指标数据，包括所有可用的资源使用情况和状态信息。
    ///
    /// ## 返回值
    ///
    /// * `Ok(serde_json::Value)` - 包含所有指标的 JSON 对象
    /// * `Err(...)` - 获取失败
    ///
    /// ## 返回数据格式
    ///
    /// ```json
    /// {
    ///   "name": "sandbox-name",
    ///   "running": true,
    ///   "cpu_usage": 12.5,
    ///   "memory_usage": 256,
    ///   "disk_usage": 1048576
    /// }
    /// ```
    ///
    /// ## 使用示例
    ///
    /// ```rust,no_run
    /// # use microsandbox_sdk::{PythonSandbox, BaseSandbox};
    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// # let sandbox = PythonSandbox::create("test").await?;
    /// # sandbox.start(None).await?;
    /// let metrics = sandbox.metrics().await?;
    /// let all = metrics.all().await?;
    ///
    /// println!("沙箱名称：{}", all["name"]);
    /// println!("运行状态：{}", all["running"]);
    /// # Ok(())
    /// # }
    /// ```
    pub async fn all(&self) -> Result<serde_json::Value, Box<dyn Error + Send + Sync>> {
        self.get_metrics().await
    }

    /// # 获取 CPU 使用率
    ///
    /// 返回沙箱的 CPU 使用率百分比。
    ///
    /// ## 返回值
    ///
    /// * `Ok(Some(f32))` - CPU 使用率百分比（0-100）
    /// * `Ok(None)` - 指标不可用（沙箱未运行或服务器未提供）
    /// * `Err(...)` - 获取失败
    ///
    /// ## 说明
    ///
    /// - `0.0` - 完全空闲
    /// - `100.0` - 完全占用
    /// - `50.0` - 一半的 CPU 资源正在使用
    ///
    /// 注意：对于 idle 的沙箱，CPU 使用率可能为 0.0 或不精确。
    ///
    /// ## 使用示例
    ///
    /// ```rust,no_run
    /// # use microsandbox_sdk::{PythonSandbox, BaseSandbox};
    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// # let sandbox = PythonSandbox::create("test").await?;
    /// # sandbox.start(None).await?;
    /// let metrics = sandbox.metrics().await?;
    ///
    /// if let Some(cpu) = metrics.cpu().await? {
    ///     println!("CPU 使用率：{:.1}%", cpu);
    /// } else {
    ///     println!("CPU 指标不可用");
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn cpu(&self) -> Result<Option<f32>, Box<dyn Error + Send + Sync>> {
        let metrics = self.get_metrics().await?;
        // as_f64() 尝试将 JSON 值转换为 f64
        // map(|v| v as f32) 将 f64 转换为 f32
        Ok(metrics
            .get("cpu_usage")
            .and_then(|v| v.as_f64())
            .map(|v| v as f32))
    }

    /// # 获取内存使用量
    ///
    /// 返回沙箱的内存使用量（以 MiB 为单位）。
    ///
    /// ## 返回值
    ///
    /// * `Ok(Some(u64))` - 内存使用量（MiB）
    /// * `Ok(None)` - 指标不可用
    /// * `Err(...)` - 获取失败
    ///
    /// ## 单位说明
    ///
    /// - 1 MiB = 1024 KiB = 1,048,576 字节
    /// - 例如：`256` 表示 256 MiB
    ///
    /// ## 使用示例
    ///
    /// ```rust,no_run
    /// # use microsandbox_sdk::{PythonSandbox, BaseSandbox};
    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// # let sandbox = PythonSandbox::create("test").await?;
    /// # sandbox.start(None).await?;
    /// let metrics = sandbox.metrics().await?;
    ///
    /// if let Some(memory) = metrics.memory().await? {
    ///     println!("内存使用：{} MiB", memory);
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn memory(&self) -> Result<Option<u64>, Box<dyn Error + Send + Sync>> {
        let metrics = self.get_metrics().await?;
        // as_u64() 尝试将 JSON 值转换为 u64
        Ok(metrics.get("memory_usage").and_then(|v| v.as_u64()))
    }

    /// # 获取磁盘使用量
    ///
    /// 返回沙箱的磁盘使用量（以字节为单位）。
    ///
    /// ## 返回值
    ///
    /// * `Ok(Some(u64))` - 磁盘使用量（字节）
    /// * `Ok(None)` - 指标不可用
    /// * `Err(...)` - 获取失败
    ///
    /// ## 单位转换
    ///
    /// ```rust
    /// # fn example(disk: u64) {
    /// // 字节 → KB
    /// let kb = disk as f64 / 1024.0;
    ///
    /// // 字节 → MB
    /// let mb = disk as f64 / 1024.0 / 1024.0;
    ///
    /// // 字节 → GB
    /// let gb = disk as f64 / 1024.0 / 1024.0 / 1024.0;
    /// # }
    /// ```
    ///
    /// ## 使用示例
    ///
    /// ```rust,no_run
    /// # use microsandbox_sdk::{PythonSandbox, BaseSandbox};
    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// # let sandbox = PythonSandbox::create("test").await?;
    /// # sandbox.start(None).await?;
    /// let metrics = sandbox.metrics().await?;
    ///
    /// if let Some(disk) = metrics.disk().await? {
    ///     println!("磁盘使用：{:.2} MB", disk as f64 / 1024.0 / 1024.0);
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn disk(&self) -> Result<Option<u64>, Box<dyn Error + Send + Sync>> {
        let metrics = self.get_metrics().await?;
        Ok(metrics.get("disk_usage").and_then(|v| v.as_u64()))
    }

    /// # 检查沙箱是否正在运行
    ///
    /// 返回沙箱的当前运行状态。
    ///
    /// ## 返回值
    ///
    /// * `Ok(true)` - 沙箱正在运行
    /// * `Ok(false)` - 沙箱已停止或未运行
    /// * `Err(...)` - 获取失败
    ///
    /// ## 与 `is_started()` 的区别
    ///
    /// - `is_started()` - 检查 SDK 内部状态（是否调用了 `start()`）
    /// - `is_running()` - 查询服务器获取实际运行状态
    ///
    /// 通常情况下两者应该一致，但如果沙箱在外部被停止（例如服务器重启），
    /// `is_running()` 会反映真实状态。
    ///
    /// ## 使用示例
    ///
    /// ```rust,no_run
    /// # use microsandbox_sdk::{PythonSandbox, BaseSandbox};
    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// # let sandbox = PythonSandbox::create("test").await?;
    /// # sandbox.start(None).await?;
    /// let metrics = sandbox.metrics().await?;
    ///
    /// if metrics.is_running().await? {
    ///     println!("沙箱正在运行，可以执行代码");
    /// } else {
    ///     println!("沙箱未运行");
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn is_running(&self) -> Result<bool, Box<dyn Error + Send + Sync>> {
        let metrics = self.get_metrics().await?;
        // 获取 running 字段，如果不存在或不是布尔值则返回 false
        Ok(metrics
            .get("running")
            .and_then(|v| v.as_bool())
            .unwrap_or(false))
    }
}

File: sdk/python/microsandbox/metrics.py
"""
指标监控模块 (Metrics)

本模块提供了获取沙箱资源使用情况的接口。

主要类：
    Metrics: 用于检索沙箱的资源指标

支持的指标类型：
    - CPU 使用率 (cpu 方法)
    - 内存使用量 (memory 方法)
    - 磁盘使用量 (disk 方法)
    - 运行状态 (is_running 方法)
    - 所有指标 (all 方法)

使用示例：
    from microsandbox import PythonSandbox

    async with PythonSandbox.create() as sandbox:
        metrics = sandbox.metrics

        # 获取单个指标
        cpu = await metrics.cpu()
        memory = await metrics.memory()

        # 获取所有指标
        all_metrics = await metrics.all()
        print(f"CPU: {all_metrics.get('cpu_usage')}%")
        print(f"内存：{all_metrics.get('memory_usage')} MiB")
"""

import uuid
from typing import Optional


class Metrics:
    """
    指标监控类 (Metrics Class)

    此类提供了获取沙箱资源使用情况的方法。

    此类通过 BaseSandbox 的 metrics 属性访问，不直接实例化。

    主要功能：
        - 获取 CPU 使用率 (cpu 方法)
        - 获取内存使用量 (memory 方法)
        - 获取磁盘使用量 (disk 方法)
        - 获取所有指标 (all 方法)
        - 检查运行状态 (is_running 方法)

    使用示例：
        sandbox = await PythonSandbox.create()
        metrics = sandbox.metrics

        # 获取 CPU 使用率
        cpu = await metrics.cpu()
        print(f"CPU 使用率：{cpu}%")

        # 持续监控
        while True:
            cpu = await metrics.cpu()
            memory = await metrics.memory()
            print(f"CPU: {cpu}%, Memory: {memory} MiB")
    """

    def __init__(self, sandbox_instance):
        """
        初始化指标实例。

        参数：
            sandbox_instance: 此指标对象所属的沙箱实例。
                通过沙箱实例访问服务器连接和配置。

        注意事项：
            此构造函数通常不直接调用，而是通过 BaseSandbox.metrics 属性访问。
        """
        self._sandbox = sandbox_instance

    async def _get_metrics(self) -> dict:
        """
        内部方法，从服务器获取当前指标。

        此方法发送 JSON-RPC 请求到 Microsandbox 服务器，
        获取指定沙箱的资源使用情况。

        返回：
            dict: 包含沙箱指标数据的字典。
                可能的键：
                - "name": 沙箱名称
                - "running": 是否正在运行
                - "cpu_usage": CPU 使用率（百分比）
                - "memory_usage": 内存使用量（MiB）
                - "disk_usage": 磁盘使用量（字节）

        异常：
            RuntimeError: 服务器请求失败。

        注意事项：
            - 这是内部方法，通常不直接调用
            - 沙箱必须先启动才能获取指标
        """
        # 检查沙箱是否已启动
        if not self._sandbox._is_started:
            raise RuntimeError("Sandbox is not started. Call start() first.")

        # 设置请求头
        headers = {"Content-Type": "application/json"}
        if self._sandbox._api_key:
            headers["Authorization"] = f"Bearer {self._sandbox._api_key}"

        # 构建 JSON-RPC 请求数据
        request_data = {
            "jsonrpc": "2.0",
            "method": "sandbox.metrics.get",
            "params": {
                "sandbox": self._sandbox._name,
            },
            "id": str(uuid.uuid4()),
        }

        try:
            # 发送 HTTP POST 请求到服务器
            async with self._sandbox._session.post(
                f"{self._sandbox._server_url}/api/v1/rpc",
                json=request_data,
                headers=headers,
            ) as response:
                # 检查 HTTP 状态码
                if response.status != 200:
                    error_text = await response.text()
                    raise RuntimeError(f"Failed to get sandbox metrics: {error_text}")

                # 解析 JSON 响应
                response_data = await response.json()

                # 检查是否有错误
                if "error" in response_data:
                    raise RuntimeError(
                        f"Failed to get sandbox metrics: {response_data['error']['message']}"
                    )

                # 获取结果数据
                result = response_data.get("result", {})
                sandboxes = result.get("sandboxes", [])

                # 期望响应中正好有一个沙箱（我们自己的）
                if not sandboxes:
                    return {}

                # 返回第一个（也是唯一一个）沙箱的数据
                return sandboxes[0]
        except Exception as e:
            raise RuntimeError(f"Failed to get sandbox metrics: {e}")

    async def all(self) -> dict:
        """
        获取当前沙箱的所有指标。

        返回：
            dict: 包含所有沙箱指标的字典：
                {
                    "name": str,              # 沙箱名称
                    "running": bool,          # 是否正在运行
                    "cpu_usage": float,       # CPU 使用率（百分比，0-100）
                    "memory_usage": int,      # 内存使用量（MiB）
                    "disk_usage": int         # 磁盘使用量（字节）
                }
                如果指标不可用，对应字段可能为 null 或不存在。

        异常：
            RuntimeError: 沙箱未启动或请求失败。

        使用示例：
            metrics = await sandbox.metrics.all()
            print(f"沙箱：{metrics['name']}")
            print(f"运行状态：{metrics['running']}")
            print(f"CPU: {metrics['cpu_usage']}%")
            print(f"内存：{metrics['memory_usage']} MiB")
        """
        return await self._get_metrics()

    async def cpu(self) -> Optional[float]:
        """
        获取当前沙箱的 CPU 使用率。

        返回：
            float, optional: CPU 使用率百分比（0-100）。
                - None: 当指标不可用时
                - 0.0: 沙箱空闲或指标不精确时可能返回

        异常：
            RuntimeError: 沙箱未启动或请求失败。

        使用示例：
            cpu = await metrics.cpu()
            if cpu is not None:
                print(f"CPU 使用率：{cpu}%")
            else:
                print("CPU 指标不可用")
        """
        metrics = await self._get_metrics()
        return metrics.get("cpu_usage")

    async def memory(self) -> Optional[int]:
        """
        获取当前沙箱的内存使用量。

        返回：
            int, optional: 内存使用量，单位 MiB。
                - None: 当指标不可用时

        异常：
            RuntimeError: 沙箱未启动或请求失败。

        使用示例：
            memory = await metrics.memory()
            if memory is not None:
                print(f"内存使用：{memory} MiB")
            else:
                print("内存指标不可用")
        """
        metrics = await self._get_metrics()
        return metrics.get("memory_usage")

    async def disk(self) -> Optional[int]:
        """
        获取当前沙箱的磁盘使用量。

        返回：
            int, optional: 磁盘使用量，单位字节。
                - None: 当指标不可用时

        异常：
            RuntimeError: 沙箱未启动或请求失败。

        使用示例：
            disk = await metrics.disk()
            if disk is not None:
                print(f"磁盘使用：{disk} 字节 ({disk / 1024 / 1024:.2f} MB)")
            else:
                print("磁盘指标不可用")
        """
        metrics = await self._get_metrics()
        return metrics.get("disk_usage")

    async def is_running(self) -> bool:
        """
        检查沙箱是否正在运行。

        返回：
            bool: True 表示沙箱正在运行，False 表示未运行。

        异常：
            RuntimeError: 请求失败。

        使用示例：
            if await metrics.is_running():
                print("沙箱正在运行")
            else:
                print("沙箱已停止")
        """
        metrics = await self._get_metrics()
        return metrics.get("running", False)
